pkg/bpm/buffer.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bpm
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"sync"
+)
+
+type blockingBuffer struct {
```

Comment:
How is it better than a simple buffer with rwlock?

Reply:
As the `cond` and channel usage is much more complicated than using a lock.

Reply:
I think single mutex or rwlock cannot make the implementation simpler.